src/org/pentaho/reporting/platform/plugin/async/AsyncReportStatusListener.java

```diff
@@ -71,37 +77,125 @@ public synchronized int getProgress() {
   }
 
   @Override
-  public synchronized void setProgress( int progress ) {
+  public synchronized void setProgress( final int progress ) {
     this.progress = progress;
   }
 
   @Override
+  public synchronized int getPage() {
+    return page;
+  }
+
+  @Override
+  public synchronized void setPage( final int page ) {
+    this.page = page;
+  }
+
+  @Override
+  public synchronized void setRow( final int row ) {
+    this.row = row;
+  }
+
+  @Override
+  public synchronized int getRow() {
+    return row;
+  }
+
+  @Override
+  public synchronized String getActivity() {
+    return activity;
+  }
+
+  @Override
+  public synchronized void setActivity( final String activity ) {
+    this.activity = activity;
+  }
+
+  @Override
   public synchronized String getMimeType() {
     return this.mimeType;
   }
 
   @Override
-  public synchronized void reportProcessingStarted( ReportProgressEvent event ) {
-    // will be implemented in another commit
+  public synchronized void reportProcessingStarted( final ReportProgressEvent event ) {
+    this.status = AsyncExecutionStatus.WORKING;
   }
 
   @Override
-  public synchronized void reportProcessingUpdate( ReportProgressEvent event ) {
-    // will be implemented in another commit
+  public synchronized void reportProcessingUpdate( final ReportProgressEvent event ) {
+    final int activity = event.getActivity();
+    if ( firstPageMode && ReportProgressEvent.GENERATING_CONTENT == activity && page > 0 ) {
+      //First page is ready here
+      this.status = AsyncExecutionStatus.CONTENT_AVAILABLE;
+    }
+    this.activity = getActivityCode( activity );
+    this.progress = (int) ReportProgressEvent.computePercentageComplete( event, true );
+    this.page = event.getPage();
+    this.row = event.getRow();
   }
 
   @Override
-  public synchronized void reportProcessingFinished( ReportProgressEvent event ) {
-    // will be implemented in another commit
+  public synchronized void reportProcessingFinished( final ReportProgressEvent event ) {
+    final int activity = event.getActivity();
+    this.activity = getActivityCode( activity );
+    this.progress = (int) ReportProgressEvent.computePercentageComplete( event, true );
+    this.page = event.getPage();
+    this.row = event.getRow();
+    this.status = AsyncExecutionStatus.FINISHED;
+  }
+
+  public synchronized void setFirstPageMode( final boolean firstPageMode ) {
+    this.firstPageMode = firstPageMode;
   }
 
   @Override
-  public String toString() {
+  public synchronized IAsyncReportState clone() {
```

Comment:
clone is back? Please do NOT use the clone() name unless you intending to uphold the contract *and* it makes sense to how you intend to use the method. Show your intend of how and where you use the method. 

You use the method to get a "ReportState", so please name it "getReportState". The fact that you make a copy of this instance is an implementation detail - a caller who just sees the public interface of your class should not need to know the details.